crates/rome_js_analyze/src/utils/react.rs

```diff
@@ -0,0 +1,240 @@
+use std::collections::{HashMap, HashSet};
+
+use rome_js_semantic::{Capture, ClosureExtensions, SemanticModel};
+use rome_js_syntax::{
+    JsAnyExpression, JsArrayBindingPattern, JsArrayBindingPatternElementList, JsCallExpression,
+    JsIdentifierBinding, JsVariableDeclarator, TextRange,
+};
+use rome_rowan::AstNode;
+use serde::{Deserialize, Serialize};
+
+/// Return result of [react_hook_with_dependency].
+pub(crate) struct ReactCallWithDependencyResult {
+    pub(crate) function_name_range: TextRange,
+    pub(crate) closure_node: Option<JsAnyExpression>,
+    pub(crate) dependencies_node: Option<JsAnyExpression>,
+}
+
+impl ReactCallWithDependencyResult {
+    /// Returns all [Reference] captured by the closure argument of the React hook.  
+    /// See [react_hook_with_dependency].
+    pub fn all_captures(&self, model: &SemanticModel) -> Vec<Capture> {
+        if let Some(closure) = self
+            .closure_node
+            .as_ref()
+            .and_then(|node| node.as_js_arrow_function_expression())
+            .map(|x| x.closure(model))
+        {
+            let range = closure.closure_range();
+
+            let mut descendents = closure.descendents();
+
+            let mut captures: Vec<Capture> = if let Some(closure) = descendents.next() {
+                closure.all_captures().collect()
+            } else {
+                vec![]
+            };
+
+            for closure in descendents {
+                for capture in closure.all_captures() {
+                    if capture.declaration_range().intersect(*range).is_none() {
+                        captures.push(capture);
+                    }
+                }
+            }
+            captures
+        } else {
+            vec![]
+        }
+    }
+
+    /// Returns all dependencies of a React hook.  
+    /// See [react_hook_with_dependency]
+    pub fn all_dependencies(&self) -> Vec<JsAnyExpression> {
+        self.dependencies_node
+            .as_ref()
+            .and_then(|x| {
+                Some(
+                    x.as_js_array_expression()?
+                        .elements()
+                        .into_iter()
+                        .filter_map(|x| x.ok()?.as_js_any_expression().cloned())
+                        .collect::<Vec<_>>(),
+                )
+            })
+            .unwrap_or_default()
+    }
+}
+
+// Get the i-th argument of a call expression
```

Comment:
```suggestion
/// Get the n-th argument of a call expression, it that argument is an expression
```
The argument is called `n`. And there's a `as_js_any_expression`, which is important to track

Reply:
Done.